### PR TITLE
fix(core): handle invalid bytecode in getUnlinkedBytecode for library-linked contracts

### LINKED ISSUE #1227: `getUnlinkedBytecode` crashes with "Bytecode is not a valid hex string" when build-info contains library-linked contracts
**Package:** `@openzeppelin/upgrades-core`
**Version:** 1.44.2 (latest stable as of March 2026)
**Companion package:** `@openzeppelin/hardhat-upgrades@3.9.1`
**Severity:** High (blocks all `deployProxy`/`deployBeacon` calls in affected projects)

---

## Summary

`getUnlinkedBytecode()` in `@openzeppelin/upgrades-core` throws `"Bytecode is not a valid hex string"` when the Hardhat build-info file contains contracts with external library link references (e.g., `LockupLib`, `ERC3643BatchLib`), even when the contract being deployed has **no library dependencies**. The error makes `deployProxy()` and `deployBeacon()` unusable for any contract whose build-info file also contains a library-linked contract.

---

## Environment

| Component | Version |
|-----------|---------|
| `@openzeppelin/upgrades-core` | 1.44.2 |
| `@openzeppelin/hardhat-upgrades` | 3.9.1 |
| `@openzeppelin/contracts-upgradeable` | 5.3.0 |
| `hardhat` | 2.28.6 |
| `solc` | 0.8.24 |
| Node.js | v24.13.0 |
| OS | Linux (WSL2, kernel 6.6.87.2) |

### Project structure

The project is a Hardhat monorepo with ~115 Solidity files. Three token template contracts use external libraries via `DELEGATECALL`:

```solidity
// AssetTokenERC3643.sol — uses two external libraries
import { LockupLib } from "../../libs/LockupLib.sol";
import { ERC3643BatchLib } from "../../libs/ERC3643BatchLib.sol";

contract AssetTokenERC3643 is ... {
    function someFn() external {
        LockupLib.doSomething($.storage, ...);   // DELEGATECALL
        ERC3643BatchLib.doBatch($.storage, ...);  // DELEGATECALL
    }
}

// AssetTokenStandard.sol, AssetTokenManaged.sol — use LockupLib only
```

A separate contract with **no library dependencies** is deployed via `upgrades.deployProxy`:

```solidity
// SimpleIdentityRegistry.sol — NO library imports
contract SimpleIdentityRegistry is UUPSUpgradeable, AccessManagedUpgradeable {
    // ... pure storage, no external libraries
}
```

### Steps to reproduce

1. Compile all contracts with Hardhat (`npx hardhat compile`). This produces a single `build-info/*.json` file containing all 105+ source contracts.

2. Call `deployProxy` on the non-library contract:

```typescript
const identityRegistry = await upgrades.deployProxy(
    await ethers.getContractFactory("SimpleIdentityRegistry"),
    [accessManagerAddress, beaconAddress],
    { kind: "uups", unsafeAllow: ["constructor"] }
);
```

3. Observe the error:

```
Error: Bytecode is not a valid hex string
    at hashBytecode (@openzeppelin/upgrades-core/src/version.ts:31:11)
    at hashBytecodeWithoutMetadata (@openzeppelin/upgrades-core/src/version.ts:47:10)
    at getVersion (@openzeppelin/upgrades-core/src/version.ts:14:24)
    at getUnlinkedBytecode (@openzeppelin/upgrades-core/src/validate/query.ts:133:33)
    at getDeployData (@openzeppelin/hardhat-upgrades/src/utils/deploy-impl.ts:46:47)
```

### Expected behavior

`deployProxy("SimpleIdentityRegistry")` should succeed since `SimpleIdentityRegistry` has no library dependencies. The presence of other library-linked contracts in the same build-info file should not affect deployment of unrelated contracts.

### Actual behavior

`deployProxy` throws `"Bytecode is not a valid hex string"` and all 207 tests that use `deployProxy` through this fixture path fail with the same error.

---

### Call chain

```
deployProxy()
  → deployProxyImpl()  (hardhat-upgrades/src/utils/deploy-impl.ts:46)
    → getUnlinkedBytecode(validations, ImplFactory.bytecode)  (upgrades-core/src/validate/query.ts)
      → for each contract with linkReferences in validation data:
          → unlinkBytecode(targetBytecode, wrongLinkReferences)  (link-refs.ts)
          → getVersion(corruptedBytecode)  (version.ts)
            → hashBytecode(corruptedBytecode)  ← THROWS
```

### The bug in `getUnlinkedBytecode`

**File:** `@openzeppelin/upgrades-core/src/validate/query.ts` (line 124-138)

```typescript
export function getUnlinkedBytecode(data: ValidationData, bytecode: string): string {
  const dataV3 = normalizeValidationData(data);

  for (const validation of dataV3.log) {
    const linkableContracts = Object.keys(validation)
      .filter(name => validation[name].linkReferences.length > 0);

    for (const name of linkableContracts) {
      const { linkReferences } = validation[name];
      const unlinkedBytecode = unlinkBytecode(bytecode, linkReferences);   // ← (1)
      const version = getVersion(unlinkedBytecode);                         // ← (2) THROWS
      if (validation[name].version?.withMetadata === version.withMetadata) { // ← (3)
        return unlinkedBytecode;
      }
    }
  }

  return bytecode;
}
```

The function iterates over **every** contract in the validation data that has `linkReferences` (step 1). For each one, it calls `unlinkBytecode()` which replaces bytes at the link reference offsets in the **target** bytecode with the library placeholder strings from the **wrong** contract.

**Concrete example:**

- Target: `SimpleIdentityRegistry` bytecode (23,536 hex chars, valid hex, no libraries)
- Link references from: `AssetTokenERC3643` (has `LockupLib` at offset ~11,700 and `ERC3643BatchLib` at offset ~8,200)

`unlinkBytecode` applies `AssetTokenERC3643`'s link references to `SimpleIdentityRegistry`'s bytecode:

```
unlinkedBytecode =
  SIR_bytecode[0..16400] +
  "__$323c7c1ce3a93c7cdd681723ab6326d963$__" +    // LockupLib placeholder injected at wrong offset
  SIR_bytecode[16440..]
```

The resulting bytecode has `__$...$__` embedded in the middle of the `SimpleIdentityRegistry` bytecode. While `hashBytecode()` does handle `__$<34hex>$__` placeholders (line 24 of `version.ts`), the `unlinkBytecode` function only replaces the **address portion** (20 bytes = 40 hex chars) at the link offset with the **full placeholder** (44 chars including `__$` and `$__`). This length mismatch (40 vs 44 chars) shifts the surrounding bytes, creating invalid hex sequences.

After the regex replacements in `hashBytecode()`, the bytecode still contains non-hex characters from the boundary corruption, and the validation regex at line 30 correctly rejects it:

```typescript
if (!/^(0x)?[0-9a-fA-F]*$/.test(bytecode)) {
    throw new Error('Bytecode is not a valid hex string');  // ← HERE
}
```

### Why this is a bug

The function's purpose is to find the correct link references for a given bytecode. It does this by trying all possible link reference sets and checking if the version matches (step 3). However, when the wrong link references produce invalid bytecode (steps 1-2), the error at step 2 prevents reaching the version check at step 3. The function should treat "invalid bytecode after unlinking" as a non-match and continue to the next candidate.

### Why this might not have been caught before

The bug only manifests when **all** of these conditions are true:

1. The project has contracts with **external library link references** (not just imports — actual `DELEGATECALL` via `library Foo { function bar() external ... }`)
2. The library-linked contracts are compiled in the **same Hardhat compilation run** as the target contract (sharing a `build-info` file)
3. The target contract is deployed via `deployProxy()`/`deployBeacon()` and has **no library dependencies** itself
4. The link reference offsets from the library-linked contract happen to fall **within the bounds** of the target contract's bytecode (checked by `unlinkBytecode`'s `(start + length) * 2 <= unlinkedBytecode.length` guard)

Projects that only use inherited libraries (not external/`DELEGATECALL`) or that have separate compilation units would not hit this.

---

## Suggested Fix

Wrap the `getVersion()` call in `getUnlinkedBytecode()` with a try/catch. When `hashBytecode()` throws on corrupted bytecode from a wrong link reference set, skip that candidate and continue:

```diff
--- a/src/validate/query.ts
+++ b/src/validate/query.ts
@@ -130,9 +130,14 @@ export function getUnlinkedBytecode(data: ValidationData, bytecode: string): str
     for (const name of linkableContracts) {
       const { linkReferences } = validation[name];
       const unlinkedBytecode = unlinkBytecode(bytecode, linkReferences);
-      const version = getVersion(unlinkedBytecode);
-      if (validation[name].version?.withMetadata === version.withMetadata) {
-        return unlinkedBytecode;
+      try {
+        const version = getVersion(unlinkedBytecode);
+        if (validation[name].version?.withMetadata === version.withMetadata) {
+          return unlinkedBytecode;
+        }
+      } catch (e) {
+        // unlinkBytecode with wrong linkReferences can produce invalid bytecode — skip
+        continue;
       }
     }
   }
```

### Why this fix is safe

1. **No false positives**: If `unlinkBytecode` produces valid bytecode from the wrong link references, the version check at `validation[name].version?.withMetadata === version.withMetadata` will still fail (different contract = different bytecode hash). The try/catch only handles the case where the bytecode is so corrupted it can't even be hashed.

2. **Correct contracts still match**: When the correct link references are applied, `unlinkBytecode` produces valid bytecode that passes `hashBytecode()` and matches the stored version hash.

3. **Fallback is unchanged**: If no link reference set matches (all throw or all fail the version check), the function returns the original bytecode unchanged — same as before.

### Alternative fix (more defensive)

Pre-validate the unlinked bytecode before calling `getVersion`:

```typescript
const unlinkedBytecode = unlinkBytecode(bytecode, linkReferences);
// Skip if unlinking produced invalid hex (wrong link references for this bytecode)
if (!/^(0x)?[0-9a-fA-F]*$/.test(unlinkedBytecode)) continue;
const version = getVersion(unlinkedBytecode);
```

This avoids the overhead of the try/catch but duplicates the validation regex from `hashBytecode()`.

---

## Workaround

Until the fix is merged upstream, the issue can be worked around with a **pnpm patch**:

```bash
pnpm patch @openzeppelin/upgrades-core@1.44.2
# Edit dist/validate/query.js — apply the try/catch fix above
pnpm patch-commit '<path-to-patch-dir>'
```

This creates a `patches/@openzeppelin__upgrades-core@1.44.2.patch` file that is automatically applied on `pnpm install`.

---

## Test Verification

With the fix applied:

| Metric | Before | After |
|--------|--------|-------|
| Tests passing | 672 | 879 |
| Tests failing | 207 | 0 |
| Cause of failures | All `"Bytecode is not a valid hex string"` in `erc3643Setup` → `deployProxy("SimpleIdentityRegistry")` | N/A |

All 207 previously-failing tests were blocked by the same root cause. No test logic was wrong — only the OZ upgrades bytecode validation was incorrectly rejecting valid deployments.

---

## Affected Files

| File | Function | Issue |
|------|----------|-------|
| `src/validate/query.ts` | `getUnlinkedBytecode()` | Uncaught exception from `getVersion()` when wrong link references corrupt bytecode |
| `src/version.ts` | `hashBytecode()` | Correctly rejects invalid hex — not the bug, but the symptom |
| `src/link-refs.ts` | `unlinkBytecode()` | Correctly applies link references — produces invalid output when given wrong references (expected behavior, but caller doesn't handle it) |

### PR DESCRIPTION
**Pull request by BhariGowda**
## Motivation
Fixes #1227

`getUnlinkedBytecode()` throws `"Bytecode is not a valid hex string"` when
a Hardhat build-info file contains contracts with external library link
references (e.g., `LockupLib`, `ERC3643BatchLib`), even when the contract
being deployed has no library dependencies.

## Root Cause
The function iterates over all contracts with `linkReferences` and applies
their link references to the target bytecode via `unlinkBytecode()`. When
wrong link references are applied, the resulting bytecode is corrupted.
`getVersion()` then throws on the invalid hex string before the version
check can determine it's not a match.

## Fix
Wrap the `getVersion()` call in a try/catch. When `unlinkBytecode()` with
wrong link references produces invalid bytecode, skip that candidate and
continue to the next one.

## Testing
Added regression test. All 588 tests pass.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved robustness of bytecode validation to handle edge cases gracefully without interrupting the validation process.

* **Tests**
  * Added regression test to verify bytecode validation handles complex contract dependency scenarios correctly.

[Review Change Stack](https://app.coderabbit.ai/change-stack/OpenZeppelin/openzeppelin-upgrades/pull/1246?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/core/src/validate/query.ts [hunk 1]
```diff
@@ -130,10 +130,15 @@ export function getUnlinkedBytecode(data: ValidationData, bytecode: string): str
...
+      try {
+        const version = getVersion(unlinkedBytecode);
+
+        if (validation[name].version?.withMetadata === version.withMetadata) {
+          return unlinkedBytecode;
+        }
+      } catch (e) {
+        // unlinkBytecode with wrong linkReferences can produce invalid bytecode — skip
+        continue;
       }
```
[the remark this answers is not in the document]
Comment (BhariGowda) [reply packages/core/src/validate/query.ts]: Fixed — broadened the try/catch to also cover unlinkBytecode() failures as suggested.
